pkg/util/steps/runner.go

```diff
@@ -0,0 +1,176 @@
+	timeoutCtx, cancel := context.WithTimeout(ctx, runner.stepTimeout)
+	defer cancel()
+
+	err := wait.PollImmediateUntil(runner.pollInterval, func() (bool, error) {
+		err := s.f(timeoutCtx)
+
+		if err != nil {
+			shouldRerun, err := runner.handler.ShouldRetryStep(timeoutCtx, err, s.f)
+			return !shouldRerun, err
+		}
+
+		return true, nil
+	}, timeoutCtx.Done())
+
+	return err
+}
+
+func (s actionStep) Name() string {
+	return log.FriendlyName(s.f)
+}
+
+// Action returns a Step suitable for performing an action.
+//
+// The Action will execute f when the step is reached, timing out with a failure
+// when more time than Runner.stepTimeout has elapsed. If f returns an error,
+// the Runner will call StepResultHandler.ShouldRetryStep to check whether it
+// should be rerun (e.g. due to transient errors/unrefreshed credentials). If it
+// is not to be rerun, the Action will fail and no subsequent Steps will be run.
+func Action(f ActionFunction) actionStep {
+	return actionStep{f}
+}
+
+type ConditionFunction func(context.Context) (bool, error)
+
+type conditionStep struct {
+	f       ConditionFunction
+	timeout time.Duration
+}
+
+// Condition returns a Step suitable for checking whether subsequent Steps can
+// be executed.
+//
+// The Condition will execute f repeatedly (every Runner.pollInterval), timing
+// out with a failure when more time than the provided timeout has elapsed
+// without f returning (true, nil). If f returns err, the Condition will fail.
+// If the Condition times out or fails, no subsequent Steps will be run.
+func Condition(f ConditionFunction, timeout time.Duration) conditionStep {
+	return conditionStep{
+		f:       f,
+		timeout: timeout,
+	}
+}
+
+// run checks whether the condition function returns true every
+// runner.pollInterval. If the condition function returns an error, it will
+// instantly fail.
+func (c conditionStep) run(ctx context.Context, runner *Runner) error {
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	// Run the function every runner.pollInterval until it returns true or times
+	// out.
+	err := wait.PollImmediateUntil(runner.pollInterval, func() (bool, error) {
+		return c.f(timeoutCtx)
+	}, timeoutCtx.Done())
+
+	return err
+}
+
+func (c conditionStep) Name() string {
+	return log.FriendlyName(c.f)
+}
+
+// WithSteps executes the provided steps in order until one fails or all steps
+// are completed.
+func (s *Runner) WithSteps(ctx context.Context, steps []Step) error {
+	for _, step := range steps {
+		switch step := step.(type) {
+		case actionStep:
+			s.log.Printf("%s: running step %s", s.name, step.Name())
+
+			err := step.run(ctx, s)
+			if err != nil {
+				err = s.handler.OnStepError(ctx, err, step)
+				return err
+			}
+
+		case conditionStep:
+			s.log.Printf("%s: waiting for %s", s.name, step.Name())
+
+			err := step.run(ctx, s)
+			if err != nil {
+				err = s.handler.OnStepError(ctx, err, step)
```

Comment:
I've noticed that a `context.Context` is now being passed to condition functions, despite the fact that none of them use it. Were you just trying to be consistent? I think if it's never used, it shouldn't be passed.

Reply:
Two reasons -- consistency, and preparation for https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20200123-client-go-ctx.md where `ctx` becomes an explicit argument in the functions the Conditions call.

Reply:
Makes sense